sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala

```diff
@@ -83,10 +88,221 @@ object ReorderJoin extends Rule[LogicalPlan] with PredicateHelper {
+   */
+  @tailrec
+  private def findEligibleStarJoinPlan(
+      joinedTables: Seq[(LogicalPlan, InnerLike)],
+      eligibleFactTables: Seq[(LogicalPlan, InnerLike)],
+      conditions: Seq[Expression]): Seq[(LogicalPlan, InnerLike)] = {
+    if (eligibleFactTables.isEmpty || joinedTables.size < 2) {
+      // There is no eligible star join.
+      Seq.empty[(LogicalPlan, InnerLike)]
+    } else {
+      // Compute the size of the eligible fact tables and sort them in descending order.
+      // An eligible fact table is a base table access with valid statistics.
+      val sortedFactTables = eligibleFactTables.map { plan =>
+        TableSize(plan, computeBaseTableSize(plan._1, conditions))
+      }.collect {
+        case t @ TableSize(_, Some(_)) => t
+      }.sortBy(_.size)(implicitly[Ordering[Option[BigInt]]].reverse)
+
+      sortedFactTables match {
+        case Nil =>
+          // There are no stats available for these plans.
+          // Return an empty plan list and fall back to the default join reordering.
+          Seq.empty[(LogicalPlan, InnerLike)]
+
+        case table1 :: table2 :: _ if table1.size == table2.size =>
+          // There are more tables with the same size. Conservatively, fall back to the
+          // default join reordering since we cannot make good plan choices.
+          Seq.empty[(LogicalPlan, InnerLike)]
+
+        case TableSize(factPlan @ (factTable, _), _) :: _ =>
+          // A fact table was found. Get the largest fact table and compute the corresponding
+          // dimension tables. Conservatively, a dimension table is assumed to be joined with
+          // the fact table using equality predicates.
+          val eligibleDimPlans = joinedTables.filterNot(_._1 eq factTable).filter { dimPlan =>
+            val dimTable = dimPlan._1
+            val refs = factTable.outputSet ++ dimTable.outputSet
+            conditions.filter(p => p.isInstanceOf[EqualTo] || p.isInstanceOf[EqualNullSafe])
+              .filterNot(canEvaluate(_, factTable))
+              .filterNot(canEvaluate(_, dimTable))
+              .exists(_.references.subsetOf(refs))
+          }
+
+          val dimTables = eligibleDimPlans.map { _._1 }
+
+          if (eligibleDimPlans.isEmpty) {
+            // This is an expanding join since there are no equality joins
+            // between the current fact table and the joined tables.
+            // Look for the next eligible star join plan.
+            findEligibleStarJoinPlan(
+              joinedTables,
+              eligibleFactTables.filterNot(_._1 eq factTable),
+              conditions)
+
+          } else if (isSelectiveStarJoin(factTable, dimTables, conditions)) {
+            // This is a selective star join and all dimensions are base table access.
+            // Compute the size of the dimensions and return the star join
+            // with the most selective dimensions joined lower in the plan.
+            val sortedDims = eligibleDimPlans.map { plan =>
```

Comment:
I agree that the order of the dimensions should be determine based on the join selectivity. Using table size is a temporary approximation. But preserving the order of the dimensions would be a too conservative approach. Based on our performance results in most cases this heuristic led to a good join ordering. In case we make a mistake, we can always switch to the default, positional join order. 

I will look into the broadcast vs shuffle join ordering and get back to you.